src/app/content/components/Page/PageComponent.tsx

```diff
@@ -75,12 +82,13 @@ export default class PageComponent extends Component<PagePropTypes> {
 
     this.scrollToTopOrHashManager(prevProps.scrollToTopOrHash, this.props.scrollToTopOrHash);
 
-    if (prevProps.page !== this.props.page) {
-      await this.postProcess();
+    if (!this.shouldUpdateHighlightManagers(prevProps, this.props, runId)) {
+      return;
     }
 
-    const highlightsAddedOrRemoved = this.highlightManager.update(prevProps.highlights, {
+    const highlightsAddedOrRemoved = await this.highlightManager.update(prevProps.highlights, {
       onSelect: this.onHighlightSelect,
+      waitForMathTypesetting: true,
```

Comment:
why is this necessary? mathjax returns a promise, we wait for those promises in the `postProcess` function. is the problem that you're not `await`ing that call on line 74?

Reply:
Changing the page is causing props to change a few times - at least 3 - first time is because `page` is set to `undefined`, the second is because it is being set to the new page and the third is because we fetch highlights

Since react is not waiting for a `componentDidUpdate` when it is async, we are starting at least 2 promises when navigating (and even more when there are search results involved)

Then, when we stop (after the third rerender), we are waiting for the two mathjax promises to resolve - but when each of them resolves, they will try to update highlighters - this is why I've added `shouldUpdateHighlightManagers` check

> why is this necessary? mathjax returns a promise, we wait for those promises in the postProcess function. is the problem that you're not awaiting that call on line 74?

This is the part that I don't really understand, but without it sometimes when navigating quickly all highlights fail to load.
I was thinking that maybe it was related to `this.processing` being only a last promise so I've changed it to that array of promises and the `await Promise.all(this.processing)` + remove a promise from this array when it resolves, but it didn't help.
I think that might be related to mathjax resolving a milliseconds before the browser renders the math, but idk

Reply:
ok so i'm starting to understand all the stuff you did re-arranging the update function and adding the counter, i think that makes sense, i still don't understand why this `waitForMathTypesetting` needs to be passed in though, shouldn't it be finished because of the `await this.processing` on line 81?